README
     __                                     _             
    / /  ___  __ _  ___  ___    /\/\   ___ | |_ _   _ ___ 
   / /  / _ \/ _` |/ _ \/ __|  /    \ / _ \| __| | | / __|
  / /__|  __/ (_| |  __/\__ \ / /\/\ \ (_) | |_| |_| \__ \
  \____/\___|\__, |\___||___/ \/    \/\___/ \__|\__,_|___/
             |___/                                        

        Andrew Ayer | Nathan Partlan | Jeffrey Pfau

              http://legesmotus.cs.brown.edu


ABOUT

Leges Motus is a 2D networked, team-based shooter set in zero gravity.


GAMEPLAY

The objective of the game is to reach the opposing team's gate and stay next to it until it opens fully.

Shoot enemy players to freeze them.  They will un-freeze a number of seconds after they are frozen.

You can move by jumping off of any obstacle, and you will jump in the direction of the mouse pointer.  You will also move due to the recoil from firing your gun.  When frozen, you will bounce off of obstacles.  If you unfreeze away from obstacles, shoot your weapon to move back towards the nearest obstacle.


INSTALLING ON WINDOWS

The Windows verion of Leges Motus is distributed as a binary.  Simply unzip the archive.  Leges Motus has no installer on Windows and simply deleted to remove it.  However, configuration files will be retained in your home directory.


INSTALLING ON MAC OS X

Simply drag the Leges Motus application into the Applications folder to install Leges Motus on Mac OS X.  To install the server, you need to use the bundled Install Command Line Tools script.


BUILDING AND INSTALLING ON *NIX

Please note that only GNU Make is supported.  If the default 'make' command is not GNU Make, please install and use it where appropriate.  Generally it is called gmake when it is not make.

1. Run './configure' in the top level directory.
   You can also use the --help option to find out all of the supported options for this script.  If you're using a Mac, make sure to specify --disable-bundle for a flat build.
   This script will check to make sure you have the required libraries installed.

2. Run 'make' to build the server and client.

3. Run 'make install' as an administrator to install the server and client.

See "ADVANCED BUILDING" below for more advanced build options.


RUNNING THE CLIENT ON *NIX

To run the client, if Leges Motus is in your path, run the 'legesmotus' command.  You can also double click on legesmotus.exe on Windows and the Leges Motus application on Mac, but you cannot specify command line options in this manner.

The client supports the following command-line options:
  -n NAME        set your player name (default: your user name)
  -t red|blue    set your team (default: auto-balanced)
  -s SERVER      set the hostname of the server (default: use the server browser)
  -p PORTNO      set the port number of the server (default: 16877)
  -w WIDTH       set the screen width, in pixels (default: 1024)
  -h HEIGHT      set the screen height, in pixels (default: 768)
  -f             run the game in fullscreen
  -?, --help     display this help, and exit
      --version  display version information and exit


REQUIREMENTS

The following libraries must be installed in order to build and run the Leges Motus client:

 SDL
 SDL_image
 SDL_ttf
 SDL_mixer (optional)
 OpenGL

All of these are bundled in the Windows and Mac OS X packages, except for OpenGL, which should already be installed on the system.  No additional libraries are needed to build or run the server.

If you do not have SDL_mixer installed and/or you would like to build without sound support, you may pass the --disable-sound option to configure.


RUNNING THE SERVER

To run the server, run the 'lmserver' command, or double click on lmserver.exe on Windows.

The server supports the following command-line options:
  -c CONFFILE    load server configuration options from the given file (default: no config file)
  -o OPT=VALUE   set the configuration option named OPT to VALUE
  -m MAPNAME     set the map name (default: alpha1)
  -p PORTNO	     set the port number to listen on (default: 16877)
  -P PASSWORD    set the admin password (default: admin mode is disabled)
  -i ADDRESS     address of interface to listen on (default: listen on all interfaces)
  -d             daemonize the server (not on Windows)
  -u USERNAME    drop privileges to given user (only super user may use) (not on Windows)
  -g GROUPNAME   drop privileges to given group (only super user may use) (not on Windows)
  -l             (local server) do not register with the meta server
  -?, --help     display this help, and exit
      --version  display version information and exit

By default, the Leges Motus server will contact the meta server so other players on the Internet can connect to and play on the server.  If you would NOT like your server to appear in server browsers of other Internet users, pass the -l ("local") option to lmserver, or set the server configuration option "register_server" to "no".

If a server is run behind a firewall, and it is registered with the meta server, the metaserver will attempt to traverse the firewall so that clients outside the firewall are able to connect to the server.  However, this technique does not work with all firewalls, and is not 100% reliable.  If you intend to host a long-running server behind a firewall, you are strongly recommended to modify your firewall settings to allow/forward UDP port 16877.


SPECIFYING THE DATA DIRECTORY

By default, Leges Motus expects to find its data in /usr/local/share/games/legesmotus.  This path can be set at compile time by passing a --prefix argument to configure.  The path can be overridden at runtime by setting the $LM_DATA_DIR environment variable to the location containing the data.  The data can be found in the 'data' directory in the top-level source directory.


CONTROLS

	space        Jump in direction of mouse pointer
	left click   Fire in direction of mouse pointer
	right click  Switch weapons
	1 through 8  Switch weapons (if available)
	t            Open chat
	y            Open team chat
	~            Open chat log
	ESC          Return to main menu / close chat console
	tab          Show the scoreboard


CLIENT COMMANDS

These commands are entered in the client's chat bar:

	/tchat MESSAGE   Send a message to your team
	/name NEWNAME    Change your name
	/team red|blue   Change teams
	/server COMMAND  Send the command to the server (see below for list of server commands)
	/quit            Quit the game
	/legal           Show legal information (copyright and warranty)
	NAME: MESSAGE    Send message to specified player only


SERVER COMMANDS

These commands are entered in the client's chat bar.  Some commands (marked as [op]) require you to authenticate first with the password specified on the command-line to the server.

	/server auth PASSWORD  Authenticate with given password
	/server teamscore      Return the score for each team
	/server teamcount      Return the number of players on each team
	/server maps           Retrieve a list of maps
	/server balance        Balance the teams [op]
	/server shakeup        Randomly reform the teams [op]
	/server reset          Reset the scores [op]
	/server map MAPNAME    Load the given map [op]
	/server newgame        Start new game [op]
	/server kick PLAYER    Kick a player [op]
	/server shutdown       Shutdown the server [op]
	/server help           Display this help

If no operator password is set, then any player connecting from the localhost (127.0.0.1) will be given operator status.  If this is undesirable behavior, then be sure to set an operator password.


ON THE INTERNET

Home Page (News, Screenshots, Wiki): http://legesmotus.endrift.com/
Project Page (Downloads):            http://sourceforge.net/projects/legesmotus/
Trac Page (Bugs):                    http://sourceforge.net/apps/trac/legesmotus/
IRC Channel:                         irc.freenode.net #legesmotus
Subversion Repository:               https://legesmotus.svn.sourceforge.net/svnroot/legesmotus

For bugs and feature requests, visit Trac.  For immediate help or to try to find a game, stop by our IRC channel.  For documentation, visit the wiki.


ORGANIZATION

The code is divided into five components, located in the following directories:

	common         Common code for both client and server
	client         The client
	server         The server
	serverscanner  A command line tool for scanning a metaserver
	metaserver     The metaserver (running a metaserver is not recommended)

The data for the program are located in the data directory, and are divided into five parts:

	data/fonts    Fonts
	data/maps     Maps
	data/sounds   Sounds
	data/sprites  Sprites
	data/weapons  Weapon configurations


ADVANCED BUILDING

By default, Leges Motus will install to /usr/local.  You may change this by specifying a --prefix argument to configure.  To install into your home directory, you could use:

	./configure --prefix=$HOME/legesmotus

By default, on Mac OS X, Leges Motus will build a .app bundle, and use frameworks to find the SDL libraries.  Leges Motus will install the bundle to the system-wide /Applications directory.  To enable a traditional, flat build, use:

	./configure --disable-bundle

Note that this requires sdl-config to be installed on your system, which it generally is not unless SDL is installed through a system such as MacPorts.

To disable building a universal binary on Mac OS X, use:

	./configure --disable-universal

To build only the server or client, use one of:

	make server
	make client

To install in a different directory (useful for making packages):

	make install DESTDIR=/path/to/package/root

Note that, by default, the version of Leges Motus that is built on an OS X computer running 10.6 (Snow Leopard) or higher will not work on older versions of Mac OS X. To fix this, you must configure Leges Motus to use an older version of GCC. This can be accomplished by running the configure script with an environment variable set, like so:

	CXX=g++-4.0 ./configure

When built with a configuration like this, it should work on 10.5, and, assuming you have the SDK for 10.4 installed, 10.4 as well.


CREDITS

Original Designers and Programmers:
	Andrew Ayer      Server, networking, support code, map handling
	Nathan Partlan   Game logic, component integration, client design and interface, physics, server scanner
	Jeffrey Pfau     Graphics code, sprites, multi-platform support, server scanner
	Danny Schneider  Maps, sound, assist with client

Thanks:
	Kenny Coane         Official project mentor
	Robert Mustacchi    Shadow mentor
	Jon Silverman       All-around great guy
	Hayden Greif-Neill  Loyal advocacy and an endless list of excellent suggestions
	All our testers in the Sun Lab at Brown University's Department of Computer Science

Public Domain Sounds From:
	http://www.a1freesoundeffects.com/
	http://www.allmusiclibrary.com/free_sound_effects.php
	http://www.imphenzia.com/soundtrack/

Creative Commons Sampling 1.0 License Sounds:
	By schluppipuppie (http://www.freesound.org/usersViewSingle.php?id=4942)
            wasserflasche25.wav (http://www.freesound.org/samplesViewSingle.php?id=13389)

	By hammerklavier (http://www.freesound.org/usersViewSingle.php?id=430462)
            MOURNFUL CADENCE-Plain.aif (http://www.freesound.org/samplesViewSingle.php?id=49287)


COPYRIGHT AND WARRANTY

Copyright 2009-2010 Andrew Ayer, Nathan Partlan, Jeffrey Pfau

Leges Motus is free and open source software.  You may redistribute it and/or
modify it under the terms of version 2, or (at your option) version 3, of the
GNU General Public License (GPL), as published by the Free Software Foundation.

Leges Motus is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the full text of the GNU General Public License for
further detail.

For a full copy of the GNU General Public License, please see the COPYING file
in the root of the source code tree.  You may also retrieve a copy from
<http://www.gnu.org/licenses/gpl-2.0.txt>, or request a copy by writing to the
Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
02111-1307  USA



CHANGES
Leges Motus Version History:
--------------------------
v. 0.4.0 - 09/06/10:
 New Features
  - Several new gun types
    - Penetration gun type: can shoot through multiple players
    - Area gun type: can hit more than one player in an area
  - Weapons can now be scrolled through using the scroll wheel
  - During the game, the in-progress game can be seen behind the menu
  - Added a command-line server scanner (not in Windows distribution)
  - New protocol version that allows for incompatible game versions to share the same protocol but not connect to each other
  - The build process now does out-of-tree builds


  Bug Fixes
   - Fixed a memory leak pertaining to text shadows
   - Minor fixes enabling better portability (e.g. to OpenBSD)


  Other Changes
   - The metaserver has moved from meta.legesmotus.beanwood.com to m.legesmotus.endrift.com
--------------------------
v. 0.3.0 - 04/24/10:
 New Features:
  - Extensible and configurable set of multiple weapons.
    - Shotgun: less effective with range, but fires five projectiles in a small spread.
    - Rifle: Freezes opponent in a single hit, but longer reload time.
    - Impact cannon: Briefly freezes opponent and exerts a strong force on them. Low damage.
    - Ability to have continuous fire guns (not yet used in default guns).
    - Guns can have inaccuracy, various cooldowns, various health effects, freeze times, etc.
    - Guns can also reduce a player's freeze time (for guns that help teammates).
  - New overlay to select a weapon, accessible with the right mouse button.
    - Also adds the ability to have other "radial menus" in the future.
  - Added weapon cooldown indicator to the HUD.
  - Weapon switch delay is configurable.
  - New active map object infrastructure, with forces, hazards, etc. All configurable.
  - New map format.
  - New scripts to make the basis for a map.
  - New sprites for map hazards and backgrounds.
  - New spawn point features: specify initial velocity and grabbing an obstacle.
  - Completely new menu and text entry framework, allowing much more flexibility and power.
  - Added options for text background, shadow, sliding, and multisampling.
  - Added option to enable vertical sync for rendering.
  - Completely new networking system, which decreases problems with packet loss.
  - Support for ack packets from client to server. Improves networking.
  - Players who join when teams are equal are assigned to the losing team.
  - Players now have recharging energy. This is also configurable.
  - Jump velocity is configurable.
  - Configurable freeze time for players who join in the middle of a round or switch teams.
  - Improved accessibility to colorblind, with brackets around names of frozen players.
  - Added warnings when game time gets low.
  - Players automatically rotate to land nicely on walls if possible.
  - Better handling of health/energy.
  - Can now have horizontal scrollbars.
  - Can remove the update nag at compile time.
  - New application icon.
  - Exceptions now show an error box on Windows.
  - Added fallback resolutions if cannot detect options automatically.
  - Try to adjust if a video mode isn't valid.


 Bug Fixes:
  - Large amounts of code cleaning in movement and firing code.
  - Fixed code for grabbing onto obstacles.
  - Muzzle flashes now appear in the proper places.
  - Improved shape/bounding box code, with better rotation, scaling, and collision detection.
  - Fixed players "unfreezing" while invisible between rounds.
  - Fixed health bar display at round start.
  - Fixed possible crash from sending team message when not in game.
  - Click sound now only plays in server browser when something useful was clicked.
  - Fixed crashes in maps without any spawn points.
  - Fixed rendering of empty strings.
  - Improved font code, with resizing.
  - Fixed jumping into obstacle you're already intersecting with.
  - Fixed handling games with no timeout.
  - Fixes and enhancements to the makefile and configure script.
  - Removed all references to SDL_net.
  - Fixed bug where switching teams caused player to spawn without a gun sprite.
  - Added check to make sure player exists to unfreeze.
  - Cleaned up sound code.
  - Fixed bug where one could jump while in the menu.
  - Fixed sprite adding and removing for server browser.
  - Improved method of creating/deleting scores for the score overlay.
  - Fixed spawning if map hasn't been received yet.
  - Fixed unnecessary registering of the template player sprite with the window.
  - Fixed layering of frozen status bar + text under the score overlay.
  - Fixed potential strange behaviour if gate update packets arrive out of order.
--------------------------
v. 0.2.2 - 07/19/09:
 Bug Fixes:
  - Fix potential client crashes when receiving packets for non-existent players.
  - Client now saves the player name when you click "Apply" in the options menu.
  - Options menu is more intuitive and more seamless.
  - Client now checks for updates when it starts, instead of when opening the server browser.
  - Server now respawns players immediately when they are forcibly switched by team auto-balancing.
--------------------------
v. 0.2.1 - 07/11/09:
 Bug Fixes:
  - Fixed displaying the chat log on top of the score overlay.
  - Fixed showing the score overlay when the input bar is shown.
  - Options now only apply themselves when "Apply" is clicked, not "Back".
  - Server no longer gives points for shooting frozen players.
  - Added a global configuration file for Windows.
--------------------------
v. 0.2.0 (Beta 2) - 07/07/09:
 New Features:
  - Persistent configuration files for client and server.
  - New map format.
  - Map obstacle tiling.
  - Maps can now be downloaded from a server.
  - New sounds and improved sound handling.
  - Servers now use UDP hole punching to allow connections through firewalls.
  - New chat log shows old chat messages.
  - Spawning improvements, auto team balancing.

client/mac.m
/*
 * mac.m
 *
 * This file is part of Leges Motus, a networked, 2D shooter set in zero gravity.
 * 
 * Copyright 2009-2010 Andrew Ayer, Nathan Partlan, Jeffrey Pfau
 * 
 * Leges Motus is free and open source software.  You may redistribute it and/or
 * modify it under the terms of version 2, or (at your option) version 3, of the
 * GNU General Public License (GPL), as published by the Free Software Foundation.
 * 
 * Leges Motus is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE.  See the full text of the GNU General Public License for
 * further detail.
 * 
 * For a full copy of the GNU General Public License, please see the COPYING file
 * in the root of the source code tree.  You may also retrieve a copy from
 * <http://www.gnu.org/licenses/gpl-2.0.txt>, or request a copy by writing to the
 * Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
 * 02111-1307  USA
 * 
 */

#import "mac.h"
#import <unistd.h>
#include "mac_bridge.h"
//#import <stdio.h>

// LMMain

@implementation LMMain

- (void)applicationDidFinishLaunching:(NSNotification*)notification {
	int status;

	NSProcessInfo* pInfo = [NSProcessInfo processInfo];
	NSArray* args = [[pInfo arguments] retain];
	
	unsigned int argc = [args count];
	char const **argv = NSZoneMalloc(NSDefaultMallocZone(), argc*sizeof(char const**));
	unsigned int i;

	//FILE* log = fopen("log","w");

	// TODO see how Unicode paths are handled
	if (argc >= 2) {
		NSString *arg1 = (NSString*)[args objectAtIndex:1];
		if ([arg1 length] > 4 && [[arg1 substringToIndex:4] isEqualToString:@"-psn"]) {
			//fprintf(log,"Launched from finder %s\n",[[[NSBundle mainBundle] bundlePath] UTF8String]);
			NSString* lpath = [[NSBundle mainBundle] bundlePath];
			lpath = [lpath stringByAppendingString:@"/Contents"];
			NSString *dataPath = [lpath stringByAppendingString:@"/Resources/data"];
			NSString *execPath = [lpath stringByAppendingString:@"/MacOS"];
			setenv("LM_DATA_DIR", [dataPath UTF8String], 0);
			setenv("LM_EXEC_DIR", [execPath UTF8String], 0);
		}
	}

	for (i = 0; i < argc; ++i) {
		NSString* arg = [args objectAtIndex:i];
		argv[i] = [arg UTF8String];
		//fprintf(log,"%s\n",argv[i]);
	}

	status = SDL_main([args count], argv);

	NSZoneFree(NSDefaultMallocZone(), argv);
	[args release];

	exit(status);
}

@end

// LMApplication

@implementation LMApplication

- (void)terminate:(id)sender {
	clean_exit();
}

@end

// misc

void toplevel_exception(const char* text) {
	NSAlert* alert = [[NSAlert alertWithMessageText:[NSString stringWithUTF8String:text] defaultButton:nil alternateButton:nil
		otherButton:nil informativeTextWithFormat:@"Please ensure that Leges Motus was installed to the primary hard drive"] retain];
	[alert runModal];
	[alert release];
}

#ifdef LM_FWBASED
#ifdef main
#undef main
#endif

int main(int argc, char *argv[]) {
    return NSApplicationMain(argc, (const char **)argv);
}

#endif
